src/LockstakeUrn.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-FileCopyrightText: © 2023 Dai Foundation <www.daifoundation.org>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.21;
+
+interface VatLike {
+    function hope(address) external;
+}
+
+interface GemLike {
+    function balanceOf(address) external view returns (uint256);
+    function approve(address, uint256) external;
+    function transfer(address, uint256) external;
+}
+
+interface StakingRewardsLike {
+    function rewardsToken() external view returns (GemLike);
+    function stake(uint256, uint16) external;
+    function withdraw(uint256) external;
+    function getReward() external;
+}
+
+contract LockstakeUrn {
+    // --- immutables ---
+
+    address immutable public engine;
+    GemLike immutable public lsmkr;
+    VatLike immutable public vat;
+
+    // --- modifiers ---
+
+    modifier isEngine {
+        require(msg.sender == engine, "LockstakeUrn/not-engine");
+        _;
+    }
+
+    // --- constructor & init ---
+
+    constructor(address vat_, address lsmkr_) {
+        engine = msg.sender;
+        vat = VatLike(vat_);
+        lsmkr = GemLike(lsmkr_);
+    }
+
+    function init() external isEngine {
+        vat.hope(msg.sender);
+        lsmkr.approve(msg.sender, type(uint256).max);
```

Comment:
A small optimization here, since you seem to care about `open` gas, would be to use `engine` instead of `msg.sender` (since `engine` is immutable) and remove the `isEngine` modifier. But it will not make a large difference.